File: leetcode-js/061-rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if (!head) {
        return null;
    }

    let len = 1;
    let node = head;

    while (node.next) {
        len += 1;
        node = node.next;
    }

    node.next = head;

    if ((k %= len)) {
        for (let i = 0; i < len - k; i += 1) {
            node = node.next;
        }
    }

    const newHead = node.next;
    node.next = null;
    return newHead;
};

File: leetcode-java/061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return head;
        }

        int len = 1;
        ListNode node = head;

        while (node.next != null) {
            len += 1;
            node = node.next;
        }

        node.next = head;

        k %= len;

        if (k != 0) {
            for (int i = 0; i < len - k; i++) {
                node = node.next;
            }
        }

        ListNode newHead = node.next;
        node.next = null;
        return newHead;
    }
}
